### DISCUSSION TITLE
Why is the _id a uuid

### DISCUSSION THREAD

**Discussion by singersbalm** [ostafen/clover#148] [General]
Why can't I define a custom _id?. I tried to define a users collection with the email as _id to improve lookup performance.
```
func isValidObjectId(id string) bool {
	_, err := uuid.FromString(id)
	return err == nil
}
```

Clover only allows uuids, but why. what's the benefit? Shouldn't that check be optional?

**Comment by ostafen** [comment]
Hey, @singersbalm, you are right, that check can be skipped.
Do you want to send a PR?

**Reply by singersbalm** [reply]
I would but I don't know with what to replace it. Should I replace the check with a string length check or should it check the env var for what kind of check it should use?

**Comment by mwmahlberg** [comment]
Forgive me, but this  smells a bit like premature optimization. I have been running benchmarks, and unless you have *several hundreds of thousands of users*, the difference between using 

```golang
_ := db.CreateIndex("users", "Email")
docs, _ := db.FindAll(query.NewQuery("users").Where(query.Field("Email").Eq("john.doe@example.com"))
```

and accessing them via their ID field is negligible.  Looking up users via index yielded the following results

```none
$ go test -bench=. -v
=== RUN   TestBoltDB
--- PASS: TestBoltDB (0.03s)
goos: darwin
goarch: arm64
pkg: <does not matter>
BenchmarkBolt10kUsers
BenchmarkBolt10kUsers-10          250646              4753 ns/op
BenchmarkBolt100kUsers
BenchmarkBolt100kUsers-10         243933              5119 ns/op
BenchmarkBolt200kUsers
BenchmarkBolt200kUsers-10         265838              4894 ns/op
BenchmarkBolt300kUsers
BenchmarkBolt300kUsers-10         241802              4884 ns/op
PASS
ok      <does not matter>        658.752s

```

Mind you - that is lookups *per second*.  Also note: Most of the 11 minutes was spent on creating the several hundred K users. The data model was simple, as we wanted to test the *lookup*, not the retrieval:

```golang
type User struct {
	Name  string
	Email string
}
```

Bottom line: It is highly unlikely that the lookup will have any significant impact when compared to actually loading the data.

From my point of view, using a non-customizable `_id` which is rather resilient against collisions and using indices to optimize queries is one of the key advantages of clover, keeping things simple. We *could* argue that for both bbolt and badger, using a BSON ObjectID would be an improvement, since it is monotonically increasing and hence better suited for prefix scans, but making the `_id` field configurable will open a whole slew of new problems.
